kernel/src/listed_log_files.rs

```diff
@@ -40,32 +40,103 @@ pub(crate) struct ListedLogFiles {
-    let start_from = log_root.join(&version_prefix)?;
-
-    Ok(storage
-        .list_from(&start_from)?
-        .map(|meta| ParsedLogPath::try_from(meta?))
-        // TODO this filters out .crc files etc which start with "." - how do we want to use these kind of files?
-        .filter_map_ok(identity)
-        .take_while(move |path_res| match path_res {
-            Ok(path) => end_version.is_none_or(|end_version| end_version >= path.version),
-            Err(_) => true,
-        }))
+    if let Some(latest_log_tail) = log_tail.last() {
+        if let Some(requested_end) = end_version {
+            if requested_end > latest_log_tail.version {
+                return Err(Error::generic(format!(
+                    "Requested end version {requested_end} is beyond the latest version of the table {}",
+                    latest_log_tail.version
+                )));
+            }
+        }
+    }
+
+    // calculate listing bounds
+    let start_version = start_version.into().unwrap_or(0);
+    let end_version = end_version.unwrap_or(Version::MAX);
+    // start_from is log path to start listing from: the log root with zero-padded start version
+    let start_from = log_root.join(&format!("{start_version:020}"))?;
+    // stop before the log_tail or at the requested end, whichever comes first
+    let log_tail_start = log_tail.first();
+    let list_end_version =
+        log_tail_start.map_or(end_version, |first| first.version.saturating_sub(1));
+
+    // if the log_tail covers the entire requested range (i.e. starts at or before start_version),
+    // we skip listing entirely. note that if we don't include this check, we will end up listing
+    // then just filtering all the files we listed.
+
+    let listed_files = log_tail_start
+        // log_tail covers the entire requested range, so no listing is required
+        .is_none_or(|tail_start| start_version < tail_start.version)
+        .then(|| -> DeltaResult<_> {
+            // NOTE: since engine APIs don't limit listing, we list from start_version and filter
+            Ok(storage
+                .list_from(&start_from)?
+                .map(|meta| ParsedLogPath::try_from(meta?))
+                // NOTE: this filters out .crc files etc which start with "." - some engines
+                // produce `.something.parquet.crc` corresponding to `something.parquet`. Kernel
+                // doesn't care about these files. Critically, note these are _different_ than
+                // normal `version.crc` files which are listed + captured normally. Additionally
+                // we likely aren't even 'seeing' these files since lexicographically the string
+                // "." comes before the string "0".
+                .filter_map_ok(identity)
+                .take_while(move |path_res| match path_res {
+                    Ok(path) => path.version <= list_end_version,
+                    Err(_) => true,
+                }))
```

Comment:
aside: in _theory_ one could use `Result::is_ok_and` here, but I don't think it actually makes the code simpler or easier to read?
```suggestion
                .take_while(move |path_res| {
                    // discard any path with too-large version; keep errors
                    !path_res.is_ok_and(|path| path.version > list_end_version)
                })
```
same problem for `matches!` macro:
```suggestion
                .take_while(move |path_res| {
                    // discard any path with too-large version; keep errors
                    !matches!(path_res, Ok(path) if path.version > list_end_version)
                })
```
The curse of boolean negation -- a hypothetical `Result::is_err_or` method might have been worth using:
```suggestion
                .take_while(move |path_res| {
                    path_res.is_err_or(|path| path.version <= list_end_version)
                })
```